core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -322,6 +324,19 @@ private Object processWithExceptionHandling(Expression expression, SymbolResolve
             }
         }
 
+        private Object processWithCaching(Expression expression, SymbolResolver resolver)
+        {
+            if (optimize && resolver instanceof NoOpSymbolResolver) {
```

Comment:
Could we make it work for `optimize == false`? Maybe have a separate cache that just doesn't cache anything in case of exception

```
                    // Certain operations like 0 / 0 or likeExpression may throw exceptions.
                    // When optimizing, do not throw the exception, but delay it until the expression is actually executed.
                    // This is to take advantage of the possibility that some other optimization removes the erroneous
                    // expression from the plan.
```
@martint is it even legal to remove `0/0` if the whole expression gets simplified?

Reply:
I think you can make it work for `optimiza == false` by just not caching result in case of an exception. Please add a test

Reply:
When no exception happens, value should be cached for `optimized = false and true`